apps/ex_wire/lib/ex_wire/struct/block_queue.ex

```diff
@@ -201,86 +161,92 @@ defmodule ExWire.Struct.BlockQueue do
+  waiting still for the parent block.
   """
-  @spec process_block_queue(t, Blocktree.t(), Chain.t(), MerklePatriciaTree.DB.db()) ::
-          {t, Blocktree.t()}
+  @spec process_block_queue(t(), Blocktree.t(), Chain.t(), Trie.t()) ::
+          {t(), Blocktree.t(), Trie.t()}
   def process_block_queue(
-        block_queue = %__MODULE__{do_validation: do_validation},
+        block_queue = %__MODULE__{},
         block_tree,
         chain,
-        db
+        trie
       ) do
-    # We can only process the next canonical block
-
+    # First get ready to process blocks
     {remaining_block_queue, blocks} = get_complete_blocks(block_queue)
-    trie = MerklePatriciaTree.Trie.new(db)
-
-    block_tree =
-      Enum.reduce(blocks, block_tree, fn block, block_tree ->
-        case Blocktree.verify_and_add_block(
-               block_tree,
-               chain,
-               block,
-               trie,
-               do_validation
-             ) do
-          :parent_not_found ->
-            _ = Logger.debug("[Block Queue] Failed to verify block due to missing parent")
-
-            block_tree
-
-          {:invalid, reasons} ->
-            _ =
-              Logger.debug(fn ->
-                "[Block Queue] Failed to verify block due to #{inspect(reasons)}"
-              end)
-
-            block_tree
-
-          {:ok, {new_block_tree, _new_trie}} ->
-            _ = Logger.debug("[Block Queue] Verified block and added to new block tree")
-
-            new_block_tree
-        end
-      end)
 
-    {remaining_block_queue, block_tree}
+    # Then recursively process them
+    do_process_blocks(blocks, remaining_block_queue, block_tree, chain, trie)
+  end
+
+  @spec do_process_blocks(list(Block.t()), t(), Blocktree.t(), Chain.t(), Trie.t()) ::
+          {t(), Blocktree.t(), Trie.t()}
+
+  defp do_process_blocks([], block_queue, block_tree, _chain, trie),
+    do: {block_queue, block_tree, trie}
+
+  defp do_process_blocks([block | rest], block_queue, block_tree, chain, trie) do
+    {new_block_tree, new_trie, new_backlog, extra_blocks} =
+      case Blocktree.verify_and_add_block(
+             block_tree,
+             chain,
+             block,
+             trie,
+             block_queue.do_validation
+           ) do
+        {:invalid, [:non_genesis_block_requires_parent]} ->
+          # Note: this is probably too slow since we see a lot of blocks without
+          #       parents and, I think, we're running the full validity check.
+
+          # :ok = Logger.debug("[Block Queue] Failed to verify block due to missing parent")
+
+          updated_backlog =
+            Map.update(
+              block_queue.backlog,
+              block.header.parent_hash,
+              [block],
+              fn blocks -> [block | blocks] end
+            )
+
+          {block_tree, trie, updated_backlog, []}
+
+        {:invalid, reasons} ->
+          :ok =
```

Comment:
I think this would fail if you raise log level, no?

Reply:
From my two seconds of investigation on a local VM, seems ok:

**config**
```elixir
config :logger,
  level: :warn
```

**iex**
```elixir
iex(2)> Logger.debug("hi")
:ok
iex(3)> Logger.warn("hi")
:ok

03:01:12.289 [warn]  hi
```

Reply:
You're right, thats how Lager used to behave 🤕